backfiller/backfiller/main.py

```diff
@@ -0,0 +1,200 @@
+
+# this is a prototype of the backfiller
+# lots about web services and the like I don't know
+# needs logging, exception handling and the like
+# also proper doc strings
+
+# when starting the backfiller and every few hours, backfill everything
+# more frequently, backfill the last couple hours
+# (last three hour directories so always at least two hours).
+
+import os
+import time
+import datetime
+import errno
+import uuid
+
+import requests
+
+import common
+
+HOUR_FMT = '%Y-%m-%dT%H'
+TIMEOUT = 5
+
+def get_nodes():
+
+	# either read a config file or query the database to get the addresses
+	# of the other nodes
+	# figure out some way that the local machine isn't in the list of returned
+	# nodes so that 
+
+	# as a prototype can just hardcode some addresses.
+	# each element in nodes is a 'protocol://host:port/' string
+
+	nodes = []
+
+	return nodes
+
+def list_local_segments(base_dir, stream, variant, hour):
+
+	# based on restreamer.list_segments
+	# could just call restreamer.list_segments but this avoids http/json
+	# overheads
+	path = os.path.join(base_dir, stream, variant, hour)
+	try:
+		local_segments = [name for name in os.listdir(path) if not name.startswith('.')]
+	except OSError as e:
+		if e.errno != errno.ENOENT:
+			raise
+		
+		local_segments = []
+
+	return local_segments
+
+def list_remote_hours(node, stream, variant, timeout=TIMEOUT):
+	
+	# just a wrapper around a call to restreamer.list_hours
+	uri = '{}/files/{}/{}'.format(node, stream, variant)
+	resp = requests.get(uri, timeout=timeout)
+	hours = resp.json()
+	return hours
+
+def list_remote_segments(node, stream, variant, hour, timeout=TIMEOUT):
+
+	# just a wrapper around a call to restreamer.list_segments
+	uri = '{}/files/{}/{}/{}'.format(node, stream, variant, hour_str)
+	resp = requests.get(uri, timeout=timeout)
+	remote_segments = resp.json()
+	return remote_segments
+
+# based on _get_segment in downloader/main
+# very basic error handling
+def get_remote_segment(base_dir, node, stream, variant, hour, missing_segment,
+			timeout=TIMEOUT):
+
+
+	path = os.path.join(base_dir, stream, variant, hour, missing_segment)
+	if os.path.exists(path):
+		return
+
+	substrs = path.split('-')
```

Comment:
This will break when the hash contains a `-`, which it often does (we're using base64 with A-Z, a-z, 0-9 and _ and -).
It will also break if stream or variant contain a -, which i'm not sure if that's possible.

I've been considering abstracting out all path-related ops into a single spot in common to ensure they stay unified. I'm still not sure if it's worth the extra indirection. In the meantime a more correct treatment would be:
```
dir_name = os.path.dirname(path)
date, duration, _ = os.path.basename(path).split('-', 2)
temp_name = "-".join([date, duration, "temp", str(uuid.uuid4())])
temp_path = os.path.join(dir_name, "{}.ts".format(temp_name))
```